File: isma-intg-lib/isma-intg-lib-rkfehlberg/src/main/java/ru/nstu/isma/intg/lib/rungeKutta/rungeKuttaFehlberg/internal/RkFehlbergStabilityIntgController.java
package ru.nstu.isma.intg.lib.rungeKutta.rungeKuttaFehlberg.internal;

import ru.nstu.isma.intg.core.methods.*;
import ru.nstu.isma.intg.api.methods.*;
import ru.nstu.isma.intg.core.methods.utils.*;

public class RkFehlbergStabilityIntgController extends BaseStabilityIntgController {
	private static final double STABILITY_INTERVAL = 3.6;

	public double getStabilityInterval() {
		return 3.6;
	}

	public double getMaxJacobiEigenValue(final IntgPoint point) {
		final int varCount = point.getStages().length;
		final double[] deltaKs = new double[varCount];
		for (int i = 0; i < varCount; ++i) {
			final double k1 = point.getStages()[i][0];
			final double k2 = point.getStages()[i][1];
			final double k3 = point.getStages()[i][2];
			if (k2 == k1) {
				deltaKs[i] = 0.0;
			} else {
				deltaKs[i] = Math.abs(32.0 * k3 - 48.0 * k2 + 16.0 * k1) / Math.abs(k2 - k1);
			}
		}
		return MathUtils.max(deltaKs) / 9;
	}
}

File: isma-intg-lib/isma-intg-lib-rkfehlberg/src/main/kotlin/ru/nstu/isma/intg/lib/rungeKutta/rungeKuttaFehlberg/internal/RkFehlbergStabilityIntgController.kt
package ru.nstu.isma.intg.lib.rungeKutta.rungeKuttaFehlberg.internal

import ru.nstu.isma.intg.api.methods.IntgPoint
import ru.nstu.isma.intg.core.methods.BaseStabilityIntgController
import ru.nstu.isma.intg.core.methods.utils.MathUtils
import kotlin.math.abs

class RkFehlbergStabilityIntgController : BaseStabilityIntgController() {
    public override fun getStabilityInterval(): Double {
        return 3.6
    }

    public override fun getMaxJacobiEigenValue(point: IntgPoint): Double {
        val varCount = point.stages.size
        val deltaKs = DoubleArray(varCount)
        for (i in 0 until varCount) {
            val k1 = point.stages[i][0]
            val k2 = point.stages[i][1]
            val k3 = point.stages[i][2]
            if (k2 == k1) {
                deltaKs[i] = 0.0
            } else {
                deltaKs[i] = abs(32.0 * k3 - 48.0 * k2 + 16.0 * k1) / abs(k2 - k1)
            }
        }
        return MathUtils.max(deltaKs) / 9
    }

    companion object {
        private const val STABILITY_INTERVAL = 3.6
    }
}
